hubspot/plugin.py

```diff
@@ -417,35 +940,253 @@ def search_companies(
+        response.raise_for_status()
+        data = response.json()
+        return _parse_company(data)
+
+
+@dataclass
+class CompanyDefaultPropertyUpdate:
+    """The data required to update a company's default properties."""
+
+    company_id: str
+    updated_properties: List[Tuple[HubSpotCompanyDefaultPropertyName, str]]
+
+
+@dataclass
+class CompanyCustomPropertyUpdate:
+    """The data required to update a company's custom properties."""
+
+    company_id: str
+    updated_properties: List[Tuple[HubSpotCompanyCustomPropertyName, str]]
+
+
+def update_companies(
+    default_properties: List[CompanyDefaultPropertyUpdate],
+    custom_properties: List[CompanyCustomPropertyUpdate],
+) -> List[str]:
+    """
+    Update multiple companies in HubSpot.
+
+    Returns:
+        Company IDs that has been updated.
+    """
+    url = "https://api.hubapi.com/crm/v3/objects/companies/batch/update"
+
+    merged_properties = defaultdict(dict)
+
+    for property_update in default_properties + custom_properties:
+        merged_properties[property_update.company_id].update(
+            {prop.name: value for prop, value in property_update.updated_properties}
         )
-        companies.append(company)
 
-    return companies
+    payload = [
+        {"id": company_id, "properties": properties}
+        for company_id, properties in merged_properties.items()
+    ]
+
+    with httpx.Client(
+        transport=AugmentedTransport(actions_v0.authenticated_request_hubspot)
+    ) as client:
+        response = client.post(url, json={"inputs": payload})
+        response.raise_for_status()
+        data = response.json()
+        return [result["id"] for result in data["results"]]
+
+
+_HUBSPOT_OBJECT_TYPE_IDS = dict(
+    CONTACTS="0-1",
+    COMPANIES="0-2",
+    DEALS="0-3",
+    TICKETS="0-5",
+    CALLS="0-48",
+    EMAILS="0-49",
+    MEETINGS="0-47",
+    NOTES="0-4",
+    TASKS="0-27",
+    PRODUCTS="0-7",
+    INVOICES="0-52",
+    LINE_ITEMS="0-8",
+    PAYMENTS="0-101",
+    QUOTES="0-14",
+    SUBSCRIPTIONS="0-69",
+    COMMUNICATIONS="0-18",
+    POSTAL_MAIL="0-116",
+    MARKETING_EVENTS="0-54",
+    FEEDBACK_SUBMISSIONS="0-19",
+)
+
+
+@dataclass
+class HubSpotObjectType:
+    """name represents the name of object in HubSpot CRM."""
+
+    name: Literal[
+        "CONTACTS",
+        "COMPANIES",
+        "DEALS",
+        "TICKETS",
+        "CALLS",
+        "EMAILS",
+        "MEETINGS",
+        "NOTES",
+        "TASKS",
+        "PRODUCTS",
+        "INVOICES",
+        "LINE_ITEMS",
+        "PAYMENTS",
+        "QUOTES",
+        "SUBSCRIPTIONS",
+        "COMMUNICATIONS",
+        "POSTAL_MAIL",
+        "MARKETING_EVENTS",
+        "FEEDBACK_SUBMISSIONS",
+    ]
+
+
+@dataclass
+class HubSpotCustomObjectType:
+    name: str
+
+
+def fetch_associated_object_ids(
+    source_object_type: Union[HubSpotObjectType, HubSpotCustomObjectType],
+    target_object_type: Union[HubSpotObjectType, HubSpotCustomObjectType],
```

Comment:
Unions are good here? (if so, this is ok)

generally, Unions as args i think are fine.

Unions for return types are bad